subsys/canbus/canopen/CO_driver.c

```diff
@@ -0,0 +1,491 @@
+}
+
+void CO_CANmodule_disable(CO_CANmodule_t *CANmodule)
+{
+	int err;
+
+	if (!CANmodule || !CANmodule->dev) {
+		return;
+	}
+
+	canopen_detach_all_rx_filters(CANmodule);
+
+	err = can_configure(CANmodule->dev, CAN_SILENT_MODE, 0);
+	if (err) {
+		LOG_ERR("failed to disable CAN interface (err %d)", err);
+	}
+}
+
+CO_ReturnError_t CO_CANrxBufferInit(CO_CANmodule_t *CANmodule, u16_t index,
+				u16_t ident, u16_t mask, bool_t rtr,
+				void *object,
+				CO_CANrxBufferCallback_t pFunct)
+{
+	struct zcan_filter filter;
+	CO_CANrx_t *buffer;
+
+	if (!CANmodule || !pFunct || (index >= CANmodule->rx_size)) {
+		LOG_ERR("failed to initialize CAN rx buffer, illegal argument");
+		CO_errorReport(CANmodule->em, CO_EM_GENERIC_SOFTWARE_ERROR,
+			       CO_EMC_SOFTWARE_INTERNAL, 0);
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	buffer = &CANmodule->rx_array[index];
+	buffer->object = object;
+	buffer->pFunct = pFunct;
+
+	filter.id_type = CAN_STANDARD_IDENTIFIER;
+	filter.std_id = ident;
+	filter.std_id_mask = mask;
+	filter.rtr = (rtr ? 1 : 0);
+	filter.rtr_mask = (rtr ? 1 : 0);
+
+	if (buffer->filter_id != CAN_NO_FREE_FILTER) {
+		can_detach(CANmodule->dev, buffer->filter_id);
+	}
+
+	buffer->filter_id = can_attach_isr(CANmodule->dev,
+					   canopen_rx_isr_callback,
+					   buffer, &filter);
+	if (buffer->filter_id == CAN_NO_FREE_FILTER) {
+		LOG_ERR("failed to attach CAN rx isr, no free filter");
+		CO_errorReport(CANmodule->em, CO_EM_MEMORY_ALLOCATION_ERROR,
+			       CO_EMC_SOFTWARE_INTERNAL, 0);
+		return CO_ERROR_OUT_OF_MEMORY;
+	}
+
+	return CO_ERROR_NO;
+}
+
+CO_CANtx_t *CO_CANtxBufferInit(CO_CANmodule_t *CANmodule, u16_t index,
+			       u16_t ident, bool_t rtr, u8_t noOfBytes,
+			       bool_t syncFlag)
+{
+	CO_CANtx_t *buffer;
+
+	if (!CANmodule || (index >= CANmodule->tx_size)) {
+		LOG_ERR("failed to initialize CAN rx buffer, illegal argument");
+		CO_errorReport(CANmodule->em, CO_EM_GENERIC_SOFTWARE_ERROR,
+			       CO_EMC_SOFTWARE_INTERNAL, 0);
+		return NULL;
+	}
+
+	buffer = &CANmodule->tx_array[index];
+	buffer->ident = ident;
+	buffer->rtr = rtr;
+	buffer->DLC = noOfBytes;
+	buffer->bufferFull = false;
+	buffer->syncFlag = syncFlag;
+
+	return buffer;
+}
+
+CO_ReturnError_t CO_CANsend(CO_CANmodule_t *CANmodule, CO_CANtx_t *buffer)
+{
+	CO_ReturnError_t ret = CO_ERROR_NO;
+	struct zcan_frame msg;
+	int err;
+
+	if (!CANmodule || !CANmodule->dev || !buffer) {
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	CO_LOCK_CAN_SEND();
+
+	if (buffer->bufferFull) {
+		if (!CANmodule->first_tx_msg) {
+			CO_errorReport(CANmodule->em, CO_EM_CAN_TX_OVERFLOW,
+				       CO_EMC_CAN_OVERRUN, buffer->ident);
+		}
+		buffer->bufferFull = false;
+		ret = CO_ERROR_TX_OVERFLOW;
+	}
+
+	msg.id_type = CAN_STANDARD_IDENTIFIER;
+	msg.std_id = buffer->ident;
+	msg.dlc = buffer->DLC;
+	msg.rtr = (buffer->rtr ? 1 : 0);
+	memcpy(msg.data, buffer->data, buffer->DLC);
+
+	err = can_send(CANmodule->dev, &msg, K_NO_WAIT, canopen_tx_isr_callback,
```

Comment:
Is there a reason for K_NO_WAIT? Is this called from an ISR?

Reply:
If there are no free TX MBs, the CANopenNode stack expects the API call to return and the CAN frame to be sent at a later time, when a TX MB becomes available (this is handled in the work queue).

Reply:
`CAN_send()` can be called from "real-time" context, e.g. from the CANopen SYNC thread and must not block.

Reply:
Blocking for as long as it takes to send the CAN frame if a mailbox is available is ok by design.  Blocking while waiting for a free mailbox is not.

CANopenNode does not try to send CAN frames from interrupt context, as far as I can tell. It does call it from high-priority threads, though.